xdem/coreg/base.py

```diff
@@ -638,98 +497,448 @@ def _postprocess_coreg_apply(
+            )
+        # Otherwise, write integer directly
+        else:
+            bin_sizes = params_fit_or_bin["bin_sizes"]
+
+    # Option 1: Run fit and save optimized function parameters
+    if fit_or_bin == "fit":
+
+        # Print if verbose
+        if verbose:
+            print(
+                "Estimating alignment along variables {} by fitting "
+                "with function {}.".format(", ".join(list(bias_vars.keys())), params_fit_or_bin["fit_func"].__name__)
+            )
+
+        results = params_fit_or_bin["fit_optimizer"](
+            f=params_fit_or_bin["fit_func"],
+            xdata=np.array([var.flatten() for var in bias_vars.values()]).squeeze(),
+            ydata=values.flatten(),
+            sigma=weights.flatten() if weights is not None else None,
+            absolute_sigma=True,
+            **kwargs,
+        )
+        df = None
+
+    # Option 2: Run binning and save dataframe of result
+    elif fit_or_bin == "bin":
+
+        if verbose:
+            print(
+                "Estimating alignment along variables {} by binning "
+                "with statistic {}.".format(
+                    ", ".join(list(bias_vars.keys())), params_fit_or_bin["bin_statistic"].__name__
+                )
+            )
+
+        df = nd_binning(
+            values=values,
+            list_var=list(bias_vars.values()),
+            list_var_names=list(bias_vars.keys()),
+            list_var_bins=bin_sizes,
+            statistics=(params_fit_or_bin["bin_statistic"], "count"),
+        )
+        results = None
+
+    # Option 3: Run binning, then fitting, and save both results
+    else:
+
+        # Print if verbose
+        if verbose:
+            print(
+                "Estimating alignment along variables {} by binning with statistic {} and then fitting "
+                "with function {}.".format(
+                    ", ".join(list(bias_vars.keys())),
+                    params_fit_or_bin["bin_statistic"].__name__,
+                    params_fit_or_bin["fit_func"].__name__,
+                )
+            )
+
+        df = nd_binning(
+            values=values,
+            list_var=list(bias_vars.values()),
+            list_var_names=list(bias_vars.keys()),
+            list_var_bins=bin_sizes,
+            statistics=(params_fit_or_bin["bin_statistic"], "count"),
+        )
+
+        # Now, we need to pass this new data to the fitting function and optimizer
+        # We use only the N-D binning estimates (maximum dimension, equal to length of variable list)
+        df_nd = df[df.nd == len(bias_vars)]
+
+        # We get the middle of bin values for variable, and statistic for the diff
+        new_vars = [pd.IntervalIndex(df_nd[var_name]).mid.values for var_name in bias_vars.keys()]
+        new_diff = df_nd[params_fit_or_bin["bin_statistic"].__name__].values
+        # TODO: pass a new sigma based on "count" and original sigma (and correlation?)?
```

Comment:
is sigma used for weighting the fit later? Of course if the bin counts are very different, it would be great to take that into account during the fit by giving different weights, but probably difficult to implement.

Reply:
Yes, I'm thinking of adding this for the coreg study (propagating uncertainties while binning to pass to the later fit), it would be one of the last thing left to implement to have error-aware coregistrations! :smiley:

Reply:
For now leaving aside